crates/napi/src/js_values/de.rs

```diff
@@ -83,14 +93,47 @@ impl<'x, 'de, 'env> serde::de::Deserializer<'x> for &'de mut De<'env> {
   where
     V: Visitor<'x>,
   {
-    visitor.visit_bytes(unsafe { FromNapiValue::from_napi_value(self.0.env, self.0.value)? })
+    match type_of!(self.0.env, self.0.value)? {
+      ValueType::Object => {
+        let js_object = unsafe { JsObject::from_raw_unchecked(self.0.env, self.0.value) };
+        if js_object.is_buffer()? {
+          return visitor
+            .visit_bytes(&unsafe { BufferSlice::from_napi_value(self.0.env, self.0.value)? });
+        } else if js_object.is_arraybuffer()? {
+          return visitor.visit_bytes(unsafe {
+            let array_buf =
+              JsArrayBuffer::from_napi_value(self.0.env, self.0.value)?.into_value()?;
+            core::slice::from_raw_parts(array_buf.data as *const u8, array_buf.len)
+          });
+        }
+        visitor.visit_bytes(unsafe { FromNapiValue::from_napi_value(self.0.env, self.0.value)? })
+      }
+      _ => unreachable!(),
+    }
   }
 
   fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value>
   where
     V: Visitor<'x>,
   {
-    visitor.visit_bytes(unsafe { FromNapiValue::from_napi_value(self.0.env, self.0.value)? })
+    match type_of!(self.0.env, self.0.value)? {
+      ValueType::Object => {
+        let js_object = unsafe { JsObject::from_raw_unchecked(self.0.env, self.0.value) };
+        if js_object.is_buffer()? {
```

Comment:
I see. Since native Node-API `napi_is_buffer(Uint8Array)` returns `true` but emnapi returns `false`, so the path here is different.